src/provider/provider_os_memory.c

```diff
@@ -641,13 +653,20 @@ static int os_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
     size_t fd_offset = 0;
 
     if (fd > 0) {
+        if (util_mutex_lock(lock_fd)) {
+            LOG_ERR("locking file size failed");
+            return -1;
+        }
+
         if (*fd_size + extended_length > max_fd_size) {
+            util_mutex_unlock(lock_fd);
             LOG_ERR("cannot grow a file size beyond %zu", max_fd_size);
             return -1;
         }
 
         fd_offset = *fd_size;
         *fd_size += extended_length;
+        util_mutex_unlock(lock_fd);
```

Comment:
What about something like this? It probably should have better performance, and we don't need extra locks. 
I'm not very proficient with atomics, so it would be good if @pbalcer or @igchor check if my proposal is valid . 
```suggestion
  do {
  size_t size = *fd_size;
  if (size + extended_length > max_fd_size) {
           LOG_ERR("cannot grow a file size beyond %zu", max_fd_size);
           return -1;
  }
  fd_offset = *fd_size;
  } while(!__sync_bool_compare_and_swap(fd_size, size, size + extended_lenght));
```

Reply:
It does not work - another thread can update `fd_size` with exactly the same value ...

Reply:
It should not be a problem - as compare_and_swap checks for old value, and them updates it to new value. So even if other tread updates this value, second thread will detect it

Reply:
Right, it should work. 
But how would you do that on Windows? I do not know any equivalent of `__sync_bool_compare_and_swap` on Windows ...

Reply:
All atomic functions are available on windows - just they have stupid names. 
https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchange64

Reply:
my 2 cents: since this isn't perf critical, I'd just leave it as is with the mutex. Atomics can be tricky to get right and make things more difficult to modify.

Reply:
@lplewa ?

Reply:
I agree with Piotr - since this is not the allocation path we should use the simplest solution

Reply:
> I agree with Piotr - since this is not the allocation path we should use the simplest solution

This is an allocation path